### DISCUSSION TITLE
Custom directives in queries/mutations

### DISCUSSION THREAD

**Discussion by Kostanos** [urql-graphql/urql#2559] [Q&A] [answered]
Hey, I'm looking a way to create a custom directives.

For example: 

```GraphQL
query Products($lang: String, $showId: Bool) @custom1(someParams: "param value 1")
{
  products(lang: $lang) {
    id @include(if: $showId)
    name @custom2(anotherParam: "param value")
  }
}
```
```javascript
directivesResolvers: {
  custom1: (query, context, etc....) => {...},
  custom2: (query, context, etc....) => {...},
}
```
the goal is to do whatever you want with those directives.

Can anyone help with a clean example of how to realize it? I couldn't find anywhere yet.

Thank you.

**Comment by JoviDeCroock** [comment, marked as answer]
You can write an exchange that does just that, we haven't implemented local state as we haven't seen many use cases that demandes local extension of a server schema.

In this case it can be a pretty [basic exchange](https://formidable.com/open-source/urql/docs/advanced/authoring-exchanges/#the-rules-of-exchanges) that we put before the `fetchExchange` that removes the local fields before forwarding to the fetch one and adds them back when returning.

What you would be doing for the logic is going through the query document you receive in the operation and remove the directive you want and act on the logic. The closest exchange you can look at is the [`populateExchange`](https://github.com/FormidableLabs/urql/blob/main/exchanges/populate/src/populateExchange.ts#L261) which looks for the `@populate` directive, removes it from the query document and performs the logic
